src/PluginCreator.ts

```diff
@@ -0,0 +1,195 @@
+            ? main as ts.LanguageService
+            : undefined
+
+        this.program = this.ls
+            ? this.ls.getProgram()
+            : (
+                typeof (main as any).getTypeChecker === 'function'
+                    ? main as ts.Program
+                    : undefined
+            )
+        
+        this.checker = this.program
+            ? this.program.getTypeChecker()
+            : (
+                typeof (main as any).getBaseTypes === 'function'
+                    ? main as ts.TypeChecker
+                    : undefined
+            )
+
+        this.opts = this.program
+            ? this.program.getCompilerOptions()
+            : (
+                typeof (main as any).getBaseTypes === 'function'
+                    ? undefined
+                    : main as ts.CompilerOptions
+            )
+    }
+
+    createPlugin(factory: PluginFactory, options: PluginConfig): TransformerPlugin {
+        const name = options.transformer || options.name
+        switch (factory.type) {
+            case 'ls':
+                if (!this.ls) throw new Error(`Plugin ${name} need a LanguageService`)
+                return factory(this.ls, options)
+
+            case 'opts':
+                if (!this.opts) throw new Error(`Plugin ${name} need a CompilerOptions`)
+                return factory(this.opts, options)
+
+            case 'checker':
+                if (!this.checker) throw new Error(`Plugin ${name} need a TypeChecker`)
+                return factory(this.checker, options)
+
+            case 'program':
+            default:
+                if (!this.program) throw new Error(`Plugin ${name} need a Program`)
+                return factory(this.program, options)
+        }
+    }
+}
+
+/**
+ * @example
+ * 
+ * new PluginCreator([
+ *   {transform: '@zerollup/ts-transform-paths', someOption: '123'},
+ *   {transform: '@zerollup/ts-transform-paths', type: 'ls', someOption: '123'},
+ *   {transform: '@zerollup/ts-transform-paths', type: 'ls', after: true, someOption: '123'}
+ * ]).createTransformers(program)
+ */
+export class PluginCreator {
+    private isOldVersion: boolean
+
+    constructor(
+        private configs: PluginConfig[],
+        versionMajorMinor?: string | void,
+        private basedir: string = process.cwd()
+    ) {
+        this.isOldVersion = compareVersions('2.9', versionMajorMinor || '2.8') < 0
+    }
+
+    private resolveFactory(transform: string): PluginFactory {
+        const modulePath = resolve.sync(transform, {basedir: this.basedir})
+        const module: PluginFactory | {default: PluginFactory} = require(modulePath)
+
+        return typeof (module as any).default === 'function'
+            ? (module as any).default
+            : module
+    }
+
+    createTransformers = (main: TransformerHost) => this._createTransformers(main)
+
+    private _createTransformers(main: TransformerHost): ts.CustomTransformers {
```

Comment:
Why do you need a private method with binded public?

Reply:
To reduce function size, created in constructor. May be it is overoptimization.